reps/2024-12-4-ray-service-incr-upgrade.md

```diff
@@ -0,0 +1,279 @@
+
+The above field can be exposed through the RayService status:
+```sh
+type RayServiceStatus struct {
+    ...
+    TargetCapacity *int32 `json:"targetCapacity,omitempty"`
+}
+```
+
+The controller can then access the `target_capacity` for the old RayCluster and upgraded RayCluster through the `ActiveServiceStatus` and `PendingServiceStatus` respectively of the `RayServiceStatuses` in the RayService CR.
+
+
+#### Key requirements:
+- If users want to use the original blue/green deployment, RayService should work regardless of whether the K8s cluster has Gateway-related CRDs
+- RayService incremental upgrade should be verified on a minimal Gateway controller implementation
+
+### Gateway API
+
+Kubernetes Gateway API is a set of resources designed to manage HTTP(S), TCP, and other network traffic for Kubernetes clusters. It provides a more flexible and extensible alternative to the traditional Ingress API, offering better support for service mesh integrations, routing policies, and multi-cluster configurations. Gateway API already has a well-defined interface for traffic splitting, so KubeRay would not need to implement backends for different Ingress controllers.
+
+Gateway APIs:
+
+| API |  Use   | Status |  K8s version  |
+| ------- | ------ | ---------- | ------------- |
+|  v1.GatewayClass | Defines a Gateway Cluster level resource | GA (v0.5+) | v1.24+ |
+|  v1.Gateway | Infrastructure that binds Listeners to a set of IP addresses | GA (v0.5+) | v1.24+ |
+|  v1.HTTPRoute | Provides a way to route HTTP requests | GA (v0.5+) | v1.24+ |
+|  v1.GRPCRoute | Specifies routing behavior of gRPC requests from a Gateway listener to an API object | GA (v1.1+) | v1.25+ |
+
+
+The responsibilities for managing the Gateway API resources associated with a RayService incremental upgrade are as follows:
+- K8s cluster admin:
+  - Gateway controller - supported controllers would be those with stable support for `HTTPRoute` and `GRPcRoute`
+  - GatewayClass - defines the controller to use
+  - Gateway CRDs
+
+- RayService controller:
+  - Gateway - created per RayService and defines the listeners and routes to use
+  - HTTPRoute - defines how HTTP traffic is routed to the two RayCluster endpoints
+  - GRPcRoute - defines how GRPc traffic is routed to the two RayCluster endpoints
+
+
+### Example Upgrade Process
+
+1. K8s cluster admin installs Gateway CRDs, Gateway Class, and Gateway controller that is compatible with their infrastructure backend.
+
+2. A user creates a RayService with a `RayServiceUpgradeType` of `IncrementalUpgrade`.
+
+3. The KubeRay controller validates the values of `maxSurgePercent`, `stepSizePercent`, and `intervalSeconds`.
+
+4. KubeRay controller checks the Serve related fields (`containerPort`, etc.) of the RayService spec and creates a Gateway object. If the controller errors when attempting to create the Gateway objects (i.e. missing CRDs, Gateway controller, etc.), it defaults to a blue/green deployment strategy and continues.
```

Comment:
https://github.com/ray-project/enhancements/pull/58#discussion_r1969242039 somehow the comment is not deleted.

```suggestion
4. KubeRay controller checks the Serve related fields (`containerPort`, etc.) of the RayService spec and creates a Gateway object.
```